windows/src/engine/keyman32/kmprocessactions.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * Description: Used for apply core processor actions to
+ *              Keyman for Windows engine.
+ */
+#include "pch.h"
+
+static BOOL processUnicodeChar(AITIP* app, const km_kbp_action_item* actionItem) {
+  if (Uni_IsSMP(actionItem->character)) {
+    app->QueueAction(QIT_CHAR, (Uni_UTF32ToSurrogate1(actionItem->character)));
+    app->QueueAction(QIT_CHAR, (Uni_UTF32ToSurrogate2(actionItem->character)));
+  }
+  else {
+    app->QueueAction(QIT_CHAR, actionItem->character);
+  }
+  return TRUE;
+}
+
+static BOOL processMarker(AITIP* app, const km_kbp_action_item* actionItem) {
+  app->QueueAction(QIT_DEADKEY, (DWORD)actionItem->marker);
+  return TRUE;
+}
+
+static BOOL processAlert(AITIP* app) {
+  app->QueueAction(QIT_BELL, 0);
+  return TRUE;
+}
+
+static BOOL processBack(AITIP* app, const km_kbp_action_item* actionItem) {
+  if (actionItem->backspace.expected_type == KM_KBP_IT_MARKER) {
+    app->QueueAction(QIT_BACK, BK_DEADKEY);
+  } else if (actionItem->backspace.expected_type == KM_KBP_IT_CHAR) {
+    app->QueueAction(QIT_BACK, 0);
+  }
```

Comment:
Note that the expected types now have their own enum values `KM_KBP_BT_*`. There is also `expected_type` `KM_KBP_BT_UNKNOWN` to handle here (same action as `KM_KBP_BT_CHAR` I think):

```suggestion
  if (actionItem->backspace.expected_type == KM_KBP_BT_MARKER) {
    app->QueueAction(QIT_BACK, BK_DEADKEY);
  } else /* actionItem->backspace.expected_type == KM_KBP_BT_CHAR, KM_KBP_BT_UNKNOWN */ {
    app->QueueAction(QIT_BACK, 0);
  }
```